pyro/infer/mcmc/hmc.py

```diff
@@ -93,6 +108,36 @@ def _get_trace(self, z):
         trace_poutine(*self._args, **self._kwargs)
         return trace_poutine.trace
 
+    @staticmethod
+    def _iter_latent_nodes(trace):
+        for name, node in sorted(trace.iter_stochastic_nodes(), key=lambda x: x[0]):
+            if not (node["fn"].has_enumerate_support or isinstance(node["fn"], _Subsample)):
+                yield (name, node)
+
+    def _compute_trace_log_prob(self, model_trace):
+        if not self._has_enumerable_sites:
+            return model_trace.log_prob_sum()
+        if self.max_iarange_nesting == float("inf"):
+            raise ValueError("Finite value required for `max_iarange_nesting` when model "
+                             "has discrete (enumerable) sites.")
+        model_trace.compute_log_prob()
+        ordering = {name: frozenset(site["cond_indep_stack"])
+                    for name, site in model_trace.nodes.items()
+                    if site["type"] == "sample"}
+
+        log_probs = defaultdict(float)
+        for name, site in model_trace.nodes.items():
+            if site["type"] == "sample":
+                log_probs[ordering[name]] = log_probs[ordering[name]] + site["log_prob"]
+
+        log_prob_sum = 0.
+        for log_prob in log_probs.values():
+            enum_dim = log_prob.dim() - self.max_iarange_nesting
+            if enum_dim > 0:
+                log_prob = log_sum_exp(log_prob.reshape(-1, *log_prob.shape[enum_dim:]), dim=0)
```

Comment:
I believe you'll need to do something fancier like `Dice.in_context()` here. At the moment, you're summing over only the local probabilities rather than all probabilities. This will fail for models with nested enumeration, e.g.
```py
@poutine.broadcast
def model(data):
    p = pyro.sample("p", dist.Uniform(0., 1.))
    y = pyro.sample("y", dist.Bernoulli(p))
    q = 0.5 + 0.25 * y
    with pyro.iarange("data", len(data)):
        z = pyro.sample("z", dist.Bernoulli(q))
        mean = 2 * z - 1
        pyro.sample("obs", dist.Normal(mean, 1.), obs=data)
```
In this case I believe you'll currently compute
```py
p(p) = p_unif(p)
     + sum(p_bern(y | p) for y in [0,1])
     + prod(sum(p_bern(zi | q(y)) for zi in [0,1] for y in [0,1])
            for i in range(len(data)))
```
which is not equal to the intended
```py
p(p) = p_unif(p)
     + sum(p_bern(y | p) * prod(sum(p_bern(zi | q(y)) for zi in [0,1])
                                for i in range(len(data)))
           for y in [0,1])
```
Could we add something like this as a unit test against a vetted `log_prob` computation?

Reply:
Thanks for this example, really helped me understand the issue with nested enumeration. I will add a test to validate the log prob computation.